packages/webcomponentsjs/webcomponents-loader.js

```diff
@@ -157,16 +157,54 @@
   }
 
   if (polyfills.length) {
+    var trustedTypesPolicyWrapper = (() => {
+      var identity = function (x) {
+        return x;
+      };
+      var trustedTypesPolicyOptions = {
+        createHTML: identity,
+        createScript: identity,
+        createScriptURL: identity,
+      };
+      var trustedTypesPolicy =
+        window.trustedTypes &&
+        window.trustedTypes.createPolicy(
+          'webcomponents-loader',
+          trustedTypesPolicyOptions
+        );
+      return trustedTypesPolicy || trustedTypesPolicyOptions;
+    })();
+
     var url;
     var polyfillFile = 'bundles/webcomponents-' + polyfills.join('-') + '.js';
 
     // Load it from the right place.
     if (window.WebComponents.root) {
-      url = window.WebComponents.root + polyfillFile;
+      var trustedTypesAreEnforced = (() => {
+        try {
+          document.createElement('input').setAttribute('oninput', '');
+        } catch (e) {
+          return true;
+        }
+        return false;
+      })();
+
+      if (
+        window.trustedTypes &&
+        trustedTypesAreEnforced &&
+        !window.trustedTypes.isScriptURL(window.WebComponents.root)
+      ) {
+        throw new Error('`WebComponents.root` must be a `TrustedScriptURL`.');
+      }
+      url = trustedTypesPolicyWrapper.createScriptURL(
+        window.WebComponents.root + polyfillFile
+      );
```

Comment:
@koto, I'm curious what you think about this w.r.t. https://github.com/w3c/webappsec-trusted-types/issues/36. The issue here is that `WebComponents.root` is arbitrary input used to build a string that I'm stamping here as a `TrustedScriptURL`, so I'd like to enforce that it is also a `TrustedScriptURL`. However, if I don't check for and enforce only at the same time as the browser, then this error would be thrown even if the browser is only reporting trusted type violations. Is there a better way to propagate enforcement like this?

Reply:
We think it's an antipattern to change the behavior only if `require-trusted-types-for` (or any CSP directive, for that matter) is enforced. The issue with that is that report-only can't be safely used to iron out all the violations, and authors risk breaking production apps the moment enforcing CSP replaces the reporting one. 

The safe (but breaking) route out of this is, I think, to introduce `WebComponents.rootAsTrustedScriptURL` (naming TBD) and use only that when `window.trustedTypes` is available. 

The less-safe (but b/w compatible) is to continue using `.root`, and just documenting that the value should not be user controlled. To my understanding the loader is only used for polyfills anyway, and these would only be loaded for older browsers, which could only get TT via a polyfill, so the practical risk is minimal here.

I would personally just go for the second option, given the constraints.